spec/features/comment_spec.rb

```diff
@@ -83,6 +83,7 @@
           created_at: 90.days.ago,
           comment: "Cool story.",
         )
+        reset_session!
```

Comment:
Rather than resetting the session which uses an internal method, I would prefer that you visit the login path to perform the actual login in a feature spec.

Reply:
How can I go about this? Should I create a feature spec to test: User A logs in --> User A visits the logout path --> User B logs in?

If so, what about this test? I'd still need to somehow log out from user, so stub_login_as user1 wouldn't fail.

If it's too much trouble to explain, feel free to push the changes you'd make. Hopefully, I'll learn from that.

Reply:
To logout from the user, you can do something like:

```ruby
visit '/settings'
click 'Logout'
```

Reply:
I actually tried doing that before and it made another feature spec fail: `rspec ./spec/features/homepage_spec.rb:7`

From what I could see, **page** does not contain the expected **story.title** value when all tests are executed `bundle exec rspec`.

1. Running all tests
 a) using `reset_session!`
   **story.title** equals **story.title.18** which is contained in page 
 b) using the logout path visit `"/settings" and click_on "Logout"`
   **story.title** equals **story.title.16** which is not contained in page 

I'm not actually sure what might be causing this, but it looks like it is due to some dependency to stories previously created other tests. Perhaps we're not tearing them down?

I'll leave the error message down below just in case anyone wants to take a look at it. 

```
Failures:

  1) Reading Homepage when logged out reading a story
     Failure/Error: expect(page).to have_content(story.title)
       expected to find text "story title 18" in "Lobsters Recent Comments Search\nLogin\n1\nstory title 16 tag1 tag2 example.com\nvia username53 just now | cached | 1 comment\nModeration Log Hats Wiki Privacy About"
     # ./spec/features/homepage_spec.rb:11:in `block (3 levels) in <top (required)>'

Finished in 40.67 seconds (files took 2.67 seconds to load)
176 examples, 1 failure

Failed examples:

rspec ./spec/features/homepage_spec.rb:7 # Reading Homepage when logged out reading a story
```

Reply:
This looks like a potential source for a lot of future headaches if we don't figure it out in this PR.
@pushcx any idea what might be causing this?

Reply:
I believe I got this down to a minimal test case:

having the following click_on will fail the homepage_spec:

```diff
diff --git a/spec/features/comment_spec.rb b/spec/features/comment_spec.rb
index fe8909f..0e37460 100644
--- a/spec/features/comment_spec.rb
+++ b/spec/features/comment_spec.rb
@@ -83,7 +83,8 @@ RSpec.feature "Commenting" do
           created_at: 90.days.ago,
           comment: "Cool story.",
         )
-        reset_session!
+        visit '/settings'
+        click_on 'Logout'
 
         stub_login_as user1
         visit "/s/#{story.short_id}"
```

with the following error, which needs to be reproduced with the entire suite:

```
$ bundle exec rspec
...................................................F............................................................................................................................

Failures:

  1) Reading Homepage when logged out reading a story
     Failure/Error: expect(page).to have_content(story.title)
       expected to find text "story title 18" in "Lobsters Recent Comments Search\nLogin\n1\nstory title 16 tag1 tag2 example.com\nvia username53 just now | cached | 1 comment\nModeration Log Hats Wiki Privacy About"
     # ./spec/features/homepage_spec.rb:9:in `block (3 levels) in <top (required)>'

Finished in 13.91 seconds (files took 1.72 seconds to load)
176 examples, 1 failure

Failed examples:

rspec ./spec/features/homepage_spec.rb:7 # Reading Homepage when logged out reading a story
```

But then changing the following in the home_controller fixes the issue:

```diff
diff --git a/app/controllers/home_controller.rb b/app/controllers/home_controller.rb
index da2d18f..c3ca659 100644
--- a/app/controllers/home_controller.rb
+++ b/app/controllers/home_controller.rb
@@ -66,9 +66,8 @@ class HomeController < ApplicationController
   end
 
   def index
-    @stories, @show_more = get_from_cache(hottest: true) {
-      paginate stories.hottest
-    }
+    @stories = Story.all
+    @show_more = false
 
     @rss_link ||= {
       :title => "RSS 2.0",
```

This change passes the entire suite. So I think it is related to caching not properly being cleared between runs, I will have to investigate how lobse.rs performs caching in order to see the best way to clear it between runs.

Reply:
Here is the final diff which works for me, I had to disable the rails cache in the test environment. I think that's what we want anyways.

```diff
diff --git a/config/environments/test.rb b/config/environments/test.rb
index 0a4a643..b059bb7 100644
--- a/config/environments/test.rb
+++ b/config/environments/test.rb
@@ -7,6 +7,9 @@ Rails.application.configure do
   # and recreated between test runs. Don't rely on the data there!
   config.cache_classes = true
 
+  # disable caching when running tests
+  config.cache_store = :null_store
+
   # Do not eager load code on boot. This avoids loading your whole application
   # just for the purpose of running a single test. If you are using a tool that
   # preloads Rails for running tests, you may have to set it to true.
diff --git a/spec/features/comment_spec.rb b/spec/features/comment_spec.rb
index fe8909f..0e37460 100644
--- a/spec/features/comment_spec.rb
+++ b/spec/features/comment_spec.rb
@@ -83,7 +83,8 @@ RSpec.feature "Commenting" do
           created_at: 90.days.ago,
           comment: "Cool story.",
         )
-        reset_session!
+        visit '/settings'
+        click_on 'Logout'
 
         stub_login_as user1
         visit "/s/#{story.short_id}"
```

Reply:
Good research! Removing the story cache use from `HomeController` is a non-starter, the site would fall over. I suspect disabling caching like this would break tests of the `Keystore` model... but we don't have any of those, so this is a solid fix for now.